Minor: check_tags option

Izzette opened the pull request:
### New option—"Include tags when determining if a post matches something from your black or white lists."
- [x] Add boolean ``check_tags`` option
    * Main changes would be made at line **456** of ``data/script.js`` in ``parseSettings(savedSettings)``.
    * These changes would provided greater flexibility on how posts are filtered and increase visibility around filtering of posts by tags.
    * If updated, the extension would keep the old behavior without user modification.  This is because the default value of ``check_tags`` is true.
- [x] Check for missing options in local settings on load 
    * Changes would be made at line **38** of ``lib/main.js`` in ``parseSettings()``.
    * This would useful if any more options are added in the future, or if (for whatever reason) an option gets removed form the local settings.
    * Checks parsed settings against the default settings.
    * If a key-value pair is missing in the parsed settings, the default pair is appended to parsed settings.
    * Upon completion of the loop, if changes were made, local settings are updated to match the corrected parsed settings.
- [x] Check for deprecated settings on load
    * Just like checking for missing options, any deprecated settings (past or future) could be removed on parsing.
    * It would not be necessary in this change, as no settings would become deprecated.  However, I don't see why it shouldn't be done.
- [x] Made minor readability changes
    * They just irked me as I was reading through the source.
- [x] Tested on Chromium, Google Chrome, and Firefox
    * I found no issues in these browsers.
- [ ] Testing on other browsers and platforms
    * I do not suspect any browser or platform incompatibility issues would be introduced with this change.

bjornstar commented:
First of all, thank you for sending a PR and taking the time to look at my code.

I'm a bit confused about your check_tags option, Tumblr Savior already does check the tags and I'm not sure why you would want to turn that off. An argument could be made for an option that **only** checks tags, but I think if you put a hashtag in your blacklist entry it would accomplish the same thing.

As far as turning options on by default, my philosophy has been to not change people's settings for them and let them turn on the options that they want, when they want to enable them. New users or installs would get the new settings default values.

There is definitely room for improvement as far as the discoverability of new settings goes. That would require quite a bit more work and I'm not sure how I can do it without annoying the users.

Izzette commented:
The option being on by default would mean that the users settings are not changed, unless they change them manually.  Actually, without the addition to main.js, their settings would be changed, because the undefined setting would evaluate false (or throw an error?).

Turning tag checking of would effectively render photo posts without comments untouched.  When I first used the extension I was unsure of (and concerned about) filtered by tags, and this provides transparency in that area.

Izzette commented:
*... tag checking off would ...

bjornstar commented:
If I understand this correctly, you want an option to disable filtering tags?

Izzette commented:
Yes

bjornstar commented:
Ok, could you make it so that the option is to ignore tags?

Izzette commented:
Yeah, that's a much better idea.

Izzette commented:
Oh, and I'm a little confused about the duplicate defaultSettings in data/defaults.js and the top of data/script.js

bjornstar commented:
`data/scripts.js` is a userscript that gets injected into the page. It cannot access other files so it needs a copy of the defaults just in case it cannot get them from the extension. I prefer it to be running with the defaults rather than simply quitting if there's a problem.

Izzette commented:
Okay, changed.  I can also add an option to only check tags, if you think that would be useful.  But, with the addition of ignore_tags, it would make more sense to add an ignore_body in addition.

Izzette commented:
wait, something happend

Izzette commented:
Okay, I fixed it

Izzette commented:
I seriously improved my changes to ``parseSettings`` in ``main.js`` such that both any garbage settings are discarded, and any missing settings are incorporated.  It is also much more elegant in implementation.

I went full circle on ignore such that there are ``ignore_header`` (correctly described as posters blog name). along with ``ignore_body``.  For example, I follow a blog titled *thereisrapeinit* (actually, they just changed their name to *isitconsensual* five minutes ago).  A blacklist item for "rape" would remove all posts by this blog, a greatly undesired behavior that actually defeats the point of such an item.  The ``ignore_header`` option fixes this issue.

Izzette commented:
Okay, sorry for this.  I might have rushed into this pull request, but I'm ready now.

Izzette commented:
![screenshot from 2015-04-11 00-57-12](https://cloud.githubusercontent.com/assets/7987724/7100398/c2d3d556-dfe5-11e4-92aa-9671a9eced42.png)

Izzette commented:
Okay, sorry for this.  I might have rushed into this pull request, but I'm ready now.

I added the options to ignore the fields header, body, and tags when checking posts.  The default values for these are all false, thus, the settings don't change when updated.  In order to conserve space and make changing these settings more obvious and streamlined, I put all three options on one line:

![screenshot from 2015-04-11 00-57-12](https://cloud.githubusercontent.com/assets/7987724/7100437/8351f9ba-dfe7-11e4-86d6-e91bd47b9e0c.png)

I also added support at ``parseSettings()`` in ``main.js`` for new settings to be added with their default values, and garbage or deprecated settings to be automatically removed.  This is of course automatically extended to any new settings or the removal of any settings.  Renaming settings (which is a bad idea anyways) would cause them to be reset to their default values, but none the less updated.

![screenshot from 2015-04-11 01-17-32](https://cloud.githubusercontent.com/assets/7987724/7100458/a3eb1ca0-dfe8-11e4-8505-73fda4fdb7d7.png)

bjornstar commented on src/lib/main.js:
```diff
@@ -25,17 +25,26 @@ function chromeMessageHandler(message, sender, sendResponse) {
+		tempSettings = JSON.stringify(parsedSettings);
+		if (JSON.stringify(checkSettings) !== tempSettings)
+			localStorage.settings = tempSettings;
+	} while(false);
```
All of this object merge logic should go into another function. Then you won't have the urge to write goto statements.

bjornstar commented on src/data/script.js:
```diff
@@ -450,16 +453,22 @@ function checkPost(post) {
+		postText += postHeader.innerHTML.replace(noTags, ' ');
+
+	if (postContent && !settings.ignore_body)
+		postText += postContent.innerHTML.replace(noTags, ' ');
```
You may not realize it, but you're making a big change to the default functionality of Tumblr Savior here.

bjornstar commented on src/data/options.html:
```diff
@@ -44,6 +44,9 @@ <h1 style="line-height:32px;height:32px;margin-top:16px;">Tumblr Savior <span id
             <input type="checkbox" id="match_words_cb" name="match_words_cb" /> <label for="match_words_cb">Match whole words only. Use * as a wildcard (ie. ip* would match ipad and iphone, but not flip or mudkips).</label>
           </div>
+          <div id="ignore_div">
+            Ignore matches in these fields: Blog name <input type="checkbox" id="ignore_header_cb" name="ignore_header_cb" /> Body <input type="checkbox" id="ignore_body_cb" name="ignore_body_cb" /> Tags <input type="checkbox" id="ignore_tags_cb" name="ignore_tags_cb" />
```
Do you see how your new settings don't match the existing settings?

![screenshot from 2015-04-11 01-26-16](https://cloud.githubusercontent.com/assets/7987724/7100488/d745bfe6-dfe9-11e4-9d05-8ca6f578534d.png)

There is a  checkbox on the left and a label on the right.

Instead of `Blog name`, please use `Header`

Izzette commented:
How is this as far as the options markup goes?
![screenshot from 2015-04-11 21-15-43](https://cloud.githubusercontent.com/assets/7987724/7104127/1563fcec-e090-11e4-9a67-c0c34fff71e2.png)
